bin/orphaned_links_checker.rb

```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env ruby
+
+require "nokogiri"
+require "date"
+require "json"
+require "net/http"
+
+DOCS_PATH = "../docs/documentation"
+SITE_URL = "https://user-guide.cloud-platform.service.justice.gov.uk/"
+
+def page_urls(root_url)
+  get_all_urls(root_url, root_url)
+    .select { |_url, url| url.slice!(SITE_URL) }
+    .values
+end
+
+# Spider a URL looking for links to other pages.
+# Return a hash: { unique_page_url => url }
+def get_all_urls(url, root_url, hash = {})
+  return if hash.has_key?(url)
+
+  doc = get_nokogiri_doc(url)
+  hash[url] = url
+  get_links(doc).map { |path| get_all_urls([root_url, path].join, root_url, hash) }
+
+  hash
+end
+
+def get_links(doc)
+  doc.css('a')
+    .map { |link| link.attributes.dig("href").value }
+    .map { |href| normalise_href(href) }.compact
+    .uniq
+end
+
+def get_nokogiri_doc(url)
+  page = fetch_url(url)
+  Nokogiri::HTML(page.body)
+end
+
+def fetch_url(url)
+  uri = URI.parse(url)
+  Net::HTTP.get_response(uri)
+end
+
+# Return a normalised href value, or nil if we don't like this href
+def normalise_href(href)
+  # ignore links which are external, or to in-page anchors
+  return nil if href[0] == "#" || ["/", "http", "mail", "/ima"].include?(href[0,4])
+
+  # Remove any trailing anchors, or "/"
+  target = href.sub(/\#.*/, "").sub(/\/$/, "")
+
+  # Ignore links which don't point to html files
+  target =~ /html$/ ? target : nil
+end
+
+############################################################
+
+# DOCUMENTATION_SITES should be something like:
+# export DOCUMENTATION_SITES="https://user-guide.cloud-platform.service.justice.gov.uk"
+# NB: When this becomes too long for a bash string, we will need to
+# use an alternative mechanism for passing the list of URLs
+sites = SITE_URL
+
+pages = sites.map { |url| page_urls(url) }.flatten
+
+files = Dir["#{DOCS_PATH}/**/*.html"].select { |f| f.slice!("../docs/") }
+
+files -= pages
+
+puts files
```

Comment:
These lines feel a little odd, because after `files -= pages`, the `files` variable is no longer what it sounds like (a list of all the files).

I'd recommend doing either
```
orphaned_files = files - pages
```
or simply
```
puts (files - pages)
```
In a case as simple as this, it really doesn't matter. But, if the code grows to become more complex, it's very easy to introduce bugs because you start of thinking "I can use `files` whenever I need a list of the files which exist", but that ceases to be true later on.

There's a (very good, IMO) tendency to use "immutable" values in a lot of code, these days. i.e. once you assign a value to a variable, you leave it alone, and if you need a modified version, you create a new variable. Languages like elixir actually make it impossible to reassign a variable once you've created it. I think this is a good habit to get into.